### PR TITLE
Features/subscription-security

### PR DESCRIPTION
**Pull request by simon-ittmann**

### CODE REVIEW

**Review by apotdevin — APPROVED** [review]

File: src/main/kotlin/de/byteleaf/companyon/auth/permission/PermissionHandler.kt [hunk 1]
```diff
@@ -1,20 +1,33 @@
...
 
     private var permissionHandlers: HashMap<PermissionType, Permission> = HashMap()
+    @Autowired
+    private lateinit var securityContextService: SecurityContextService
 
     init {
         permissions.forEach { permissionHandler -> permissionHandlers.put(permissionHandler.getPermissionType(),  permissionHandler) }
     }
 
+    fun hasPermissions(permissions: List<Pair<PermissionType, String?>>? = emptyList(), skipError: Boolean = false): Boolean {
```
Comment (sauterle-byteleaf) [inline src/main/kotlin/de/byteleaf/companyon/auth/permission/PermissionHandler.kt:20]: Is skipError is unused?

File: src/main/kotlin/de/byteleaf/companyon/auth/permission/handler/CurrentUserOrAdminPermission.kt [hunk 2]
```diff
@@ -16,8 +16,12 @@ class CurrentUserOrAdminPermission : Permission() {
...
      */
     override fun hasPermission(id: String?, skipError: Boolean): Boolean {
-        if(adminPermission.hasPermission(id, id != null)) return true
-        return if(id.equals(securityContextService.getCurrentUser().id!!)) true
+        val currentUser = securityContextService.getCurrentUser()
+        if(id == null || currentUser.admin) {
+            return adminPermission.hasPermission(id, skipError)
+        }
+
+        return if(id.equals(currentUser.id!!)) true
```
Comment (sauterle-byteleaf) [inline src/main/kotlin/de/byteleaf/companyon/auth/permission/handler/CurrentUserOrAdminPermission.kt:24]: I think "!!" is not a good code style - will bring you to the same Problem in Java with NPE (NullpointerException)

File: src/main/kotlin/de/byteleaf/companyon/auth/permission/handler/CurrentUserOrAdminPermission.kt [hunk 3]
```diff
@@ -16,8 +16,12 @@ class CurrentUserOrAdminPermission : Permission() {
      * @param id if null the user must be admin! If not null the id must be the id of the current user.
      */
     override fun hasPermission(id: String?, skipError: Boolean): Boolean {
-        if(adminPermission.hasPermission(id, id != null)) return true
-        return if(id.equals(securityContextService.getCurrentUser().id!!)) true
+        val currentUser = securityContextService.getCurrentUser()
+        if(id == null || currentUser.admin) {
```
Comment (sauterle-byteleaf) [inline src/main/kotlin/de/byteleaf/companyon/auth/permission/handler/CurrentUserOrAdminPermission.kt:20]: FORMATTING: Space between "if ("

File: src/main/kotlin/de/byteleaf/companyon/user/access/UserAccessService.kt [hunk 4]
```diff
@@ -35,5 +36,7 @@ class UserAccessService {
     @IsAdmin
     fun delete(id: String): User = userService.delete(id)
 
-    fun getPublisher(): Publisher<UserUpdate> = userService.getPublisher()
+    fun getPublisher(): Publisher<UserUpdate> = userService.getPublisher { permissionHandler, event ->
+        permissionHandler.hasPermission(PermissionType.CURRENT_USER_OR_ADMIN, event.entity!!.id, true)
```
Comment (sauterle-byteleaf) [inline src/main/kotlin/de/byteleaf/companyon/user/access/UserAccessService.kt:40]: !! - Will bring you to the same Problem in Java with NPE (NullpointerException)

File: src/test/kotlin/de/byteleaf/companyon/timelog/dto/TimeLogGQLResponse.kt [hunk 5]
```diff
@@ -1,8 +1,6 @@
 package de.byteleaf.companyon.project.dto
 
-import de.byteleaf.companyon.user.dto.User
 import de.byteleaf.companyon.user.dto.UserGQLResponse
-import java.time.OffsetDateTime
```
Comment (sauterle-byteleaf) [inline src/test/kotlin/de/byteleaf/companyon/timelog/dto/TimeLogGQLResponse.kt:5]: interesting - do we need a Unused import check? ;)

**Review by sauterle-byteleaf — COMMENTED** [review]
Please check that you have a saveaction with "optimize imports" and "reformat file" .

I think it was good, when we activate a kotlin checkstyle / linting?
https://ktlint.github.io/
https://www.sonarsource.com/kotlin/


And Avoid "!!":
https://stackoverflow.com/questions/53861126/is-it-very-bad-practice-using-in-kotlin
https://phauer.com/2017/idiomatic-kotlin-best-practices/#avoid-not-null-assertions-

File: src/main/kotlin/de/byteleaf/companyon/auth/permission/handler/CurrentUserOrAdminPermission.kt (hunk 2 above)
Comment (simon-ittmann) [reply src/main/kotlin/de/byteleaf/companyon/auth/permission/handler/CurrentUserOrAdminPermission.kt:24]: I totally agree, but its not possible to have the id in the AbstractDto as required. There are some cases where the id will be set after the object initialization. If you know how to solve this issue let me know :)

File: src/main/kotlin/de/byteleaf/companyon/auth/permission/PermissionHandler.kt (hunk 1 above)
Comment (simon-ittmann) [reply src/main/kotlin/de/byteleaf/companyon/auth/permission/PermissionHandler.kt:20]: No is used for subscriptions

**Pull request merged**